lib/puppet/indirector/rest.rb

```diff
@@ -34,12 +34,44 @@ def self.srv_service
     @srv_service || :puppet
   end
 
+  # The logic for server and port is kind of gross. In summary:
+  # IF an endpoint-specific setting is requested AND that setting has been set by the user
+  #    Use that setting.
+  #         The defaults for these settings are the "normal" server/masterport settings, so
+  #         when they are unset we instead want to "fall back" to the failover-selected
+  #         host/port pair.
+  # ELSE IF we have a failover-selected host/port
+  #    Use what the failover logic came up with
+  # ELSE
+  #    Go for the legacy server/masterport settings, and hope for the best
   def self.server
-    Puppet.settings[server_setting || :server]
+    setting = server_setting()
+    if setting && setting != :server && Puppet.settings.set_by_config?(setting)
+      Puppet.settings[setting]
+    else
+      begin
+        Puppet.lookup(:server)
+      rescue
+        Puppet.debug "Dynamically-bound server lookup failed, falling back to #{setting} setting"
+        Puppet.settings[setting || :server]
+      end
+    end
   end
 
   def self.port
-    Puppet.settings[port_setting || :masterport].to_i
+    setting = port_setting()
+    srv_setting = server_setting()
+    if (setting && setting != :masterport && Puppet.settings.set_by_config?(setting))
+        || (srv_setting && srv_setting != :server && Puppet.settings.set_by_config?(srv_setting))
```

Comment:
Latest push adds this change to the default port selection logic. This now says "If this indirection has a custom server set, but no custom port set, we should use the related port setting EVEN IF it's unset" We were encountering an issue in our testing where we'd set `--server_list=garbage:123 --report_server=localhost` , and we were failing to submit reports to the working report server since the previous logic here said "`report_port` is unset, so I'll use what came from failover"